lib/queuePopulator/IngestionPopulator.js

```diff
@@ -0,0 +1,364 @@
+            if (index.queuePopulatorExtension) {
+                // eslint-disable-next-line new-cap
+                const ext = new index.queuePopulatorExtension({
+                    config: extConfig,
+                    logger: this.log,
+                });
+                ext.setZkConfig(this.zkConfig);
+                this.log.info(`${index.name} extension is active`);
+                this._extensions.push(ext);
+            }
+        });
+    }
+
+    _setupExtensions(cb) {
+        return async.each(this._extensions, (ext, next) => {
+            ext.setupZookeeper(err => {
+                if (err) {
+                    return next(err);
+                }
+                if (ext.createZkPath) {
+                    if (ext instanceof IngestionQueuePopulator) {
+                        this.extConfigs.ingestion.sources.forEach(source => {
+                            ext.createZkPath(next, source);
+                        });
+                    } else {
+                        return ext.createZkPath(next);
+                    }
+                }
+                return next();
+            });
+        }, cb);
+    }
+
+    _setupUpdatedReaders(done) {
+        const newReaders = this.logReadersUpdate;
+        this.logReadersUpdate = [];
+        async.each(newReaders, (logReader, cb) => logReader.setup(cb),
+                   err => {
+                       if (err) {
+                           return done(err);
+                       }
+                       this.logReaders = newReaders;
+                       return done();
+                   });
+    }
+
+    _closeLogState(done) {
+        if (this.raftIdDispatcher !== undefined) {
+            return this.raftIdDispatcher.unsubscribe(done);
+        }
+        return process.nextTick(done);
+    }
+
+    _processAllLogEntries(params, done) {
+        return async.map(
+            this.logReaders,
+            (logReader, done) => logReader.processAllLogEntries(params, done),
+            (err, results) => {
+                if (err) {
+                    return done(err);
+                }
+                const annotatedResults = results.map(
+                    (result, i) => Object.assign(result, {
+                        logSource: this.logReaders[i].getLogInfo(),
+                        logOffset: this.logReaders[i].getLogOffset(),
+                    }));
+                return done(null, annotatedResults);
+            });
+    }
+
+    processAllLogEntries(params, done) {
+        if (this.logReadersUpdate.length >= 1) {
+            return this._setupUpdatedReaders(err => {
+                if (err) {
+                    return done(err);
+                }
+                return this._processAllLogEntries(params, done);
+            });
+        }
+        return this._processAllLogEntries(params, done);
+    }
+
+    zkStatus() {
+        return this.zkClient.getState();
```

Comment:
since `zkClient` can be `null` before setup it would be useful to check it first.